File: src/ros_apps/carla_telemetry_cpp/src/sensors/battery.cpp
#include "carla_telemetry/sensors/battery.hpp"

#include <algorithm>
#include <cmath>

namespace carla_telemetry {

CarlaBattery::CarlaBattery(const std::string& consumption_mode, double voltage,
                           double e0, double e1, double capacity,
                           double initial_charge, double resistance, double tau,
                           double base_power_load, double power_per_speed,
                           bool start_draining, bool enable_recharge,
                           double charging_time, double update_rate,
                           double ambient_temperature)
    : v_nom_(voltage),
      e0_(e0),
      e1_(e1),
      capacity_(capacity),
      charge_(initial_charge),
      resistance_(resistance),
      tau_(tau),
      base_power_load_(base_power_load),
      power_per_speed_(power_per_speed),
      charging_time_(charging_time),
      consumption_mode_(consumption_mode),
      is_draining_(start_draining),
      ambient_temperature_(ambient_temperature),
      current_temperature_(ambient_temperature) {
  (void)enable_recharge;
  (void)update_rate;

  if (v_nom_ > 0.0 && e0_ > 0.0 && std::abs(v_nom_ - e0_) > 0.1) {
    double scale = v_nom_ / e0_;
    e0_ *= scale;
    e1_ *= scale;
  }

  state_ = make_state();
}

BatteryState CarlaBattery::update(double speed_ms, double dt) {
  if (is_depleted_ && !is_charging_) return state_;

  // Charging path
  if (is_charging_) {
    double charge_rate = capacity_ / std::max(charging_time_, 0.001);  // Ah/h

    // Update temperature (charge_rate Ah/h is numerically equal to current in
    // Amps)
    double heat_generated = charge_rate * charge_rate * resistance_;
    double heat_dissipated =
        heat_dissipation_rate_ * (current_temperature_ - ambient_temperature_);
    current_temperature_ +=
        ((heat_generated - heat_dissipated) / thermal_mass_) * dt;

    charge_ = std::min(capacity_, charge_ + charge_rate * (dt / 3600.0));
    if (charge_ >= capacity_) {
      is_charging_ = false;
      is_depleted_ = false;
    }
    state_ = make_state();
    return state_;
  }

  // Discharging path
  if (!is_draining_) return state_;

  double total_power;
  if (consumption_mode_ == "velocity_based") {
    total_power = base_power_load_ + power_per_speed_ * speed_ms;
  } else {
    total_power = base_power_load_;
  }

  double v_oc = e0_ + e1_ * (1.0 - charge_ / capacity_);
  v_oc = std::max(v_oc, 0.01);

  double i_instant = total_power / v_oc;

  if (tau_ > 0.0) {
    smoothed_current_ += (dt / tau_) * (i_instant - smoothed_current_);
  } else {
    smoothed_current_ = i_instant;
  }

  // Update temperature
  double heat_generated = smoothed_current_ * smoothed_current_ * resistance_;
  double heat_dissipated =
      heat_dissipation_rate_ * (current_temperature_ - ambient_temperature_);
  current_temperature_ +=
      ((heat_generated - heat_dissipated) / thermal_mass_) * dt;

  charge_ -= smoothed_current_ * (dt / 3600.0);
  charge_ = std::max(charge_, 0.0);

  if (charge_ <= 0.0) {
    is_depleted_ = true;
    is_draining_ = false;
  }

  state_ = make_state(total_power, v_oc);
  return state_;
}

void CarlaBattery::start_draining() {
  is_depleted_ = false;
  is_draining_ = true;
  is_charging_ = false;
}

void CarlaBattery::stop_draining() { is_draining_ = false; }

void CarlaBattery::start_charging() {
  is_charging_ = true;
  is_draining_ = false;
}

void CarlaBattery::stop_charging() { is_charging_ = false; }

BatteryState CarlaBattery::make_state(double total_power, double v_oc) const {
  double oc = v_oc;
  if (oc < 0.0) {
    oc = e0_ + e1_ * (1.0 - charge_ / capacity_);
  }
  double v_terminal = oc - smoothed_current_ * resistance_;
  double frac = std::clamp(charge_ / capacity_, 0.0, 1.0);

  BatteryState s;
  s.voltage = v_terminal;
  s.open_circuit_v = oc;
  s.current = smoothed_current_;
  s.charge = charge_;
  s.capacity = capacity_;
  s.charge_fraction = frac;
  s.percentage = frac * 100.0;
  s.power_load = total_power;
  s.temperature = current_temperature_;
  s.is_depleted = is_depleted_;
  return s;
}

}  // namespace carla_telemetry

File: src/ros_apps/carla_telemetry/carla_telemetry/sensors/battery.py
"""
| File: battery.py
| Description: Direct port of Micropolis.Telemetry.sensors.battery.Battery.
               Software battery simulation model — no CARLA sensor required.
               Driven by the ego vehicle speed obtained from vehicle.get_velocity().

Physics reference:
  https://gazebosim.org/api/sim/7/battery.html
  gz-sim/src/systems/linear_battery_plugin/LinearBatteryPlugin.cc

Two consumption modes (selectable in carla_interface_config.yaml):
  "constant"       - fixed 'power_load' watts regardless of motion
  "velocity_based" - baseline + extra power proportional to vehicle speed
"""
__all__ = ["CarlaBattery"]

import time
import numpy as np


class CarlaBattery:
    """Gazebo-compatible linear battery simulation for CARLA.

    Identical logic to ``Micropolis.Telemetry.sensors.battery.Battery``.

    Output state dict keys:
    ===================== =====================================================
    Key                   Value
    ===================== =====================================================
    ``voltage``           Terminal voltage (V)
    ``open_circuit_v``    Open-circuit voltage (V)
    ``current``           Smoothed draw current (A)
    ``charge``            Remaining charge (Ah)
    ``charge_fraction``   Remaining charge 0 – 1
    ``percentage``        Remaining charge 0 – 100 %
    ``power_load``        Instantaneous total power draw (W)
    ``is_depleted``       True when charge ≤ 0
    ===================== =====================================================
    """

    def __init__(self, config: dict = None):
        if config is None:
            config = {}
        """
        Args:
            config: Battery config dict from carla_interface_config.yaml (all keys optional).
        """
        self._update_rate = float(config.get("update_rate", 10.0))

        # ── Electrical parameters ─────────────────────────────────────
        self._v_nom      = float(config.get("voltage", 12.6))
        self._e0         = float(config.get("open_circuit_voltage_constant_coef", 12.6))
        self._e1         = float(config.get("open_circuit_voltage_linear_coef", -2.8))
        self._capacity   = float(config.get("capacity", 10.0))
        self._charge     = float(config.get("initial_charge", self._capacity))
        self._resistance = float(config.get("resistance", 0.061))
        self._tau        = float(config.get("smooth_current_tau", 2.0))

        # ── Power consumers ───────────────────────────────────────────
        self._base_power_load = float(config.get("power_load", 50.0))
        self._start_draining  = bool(config.get("start_draining", True))
        self._is_draining     = self._start_draining
        self._enable_recharge = bool(config.get("enable_recharge", False))
        self._charging_time   = float(config.get("charging_time", 1.0))   # hours
        self._is_charging     = False

        # ── Consumption mode ──────────────────────────────────────────
        self._consumption_mode = config.get("consumption_mode", "velocity_based")
        self._power_per_speed  = float(config.get("power_per_speed", 15.0))

        # ── Runtime state ─────────────────────────────────────────────
        self._smoothed_current = 0.0
        self._is_depleted      = False
        self._state = self._make_state()

    # ------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------

    @property
    def state(self):
        return self._state

    @property
    def is_depleted(self) -> bool:
        return self._is_depleted

    @property
    def is_draining(self) -> bool:
        return self._is_draining

    # ------------------------------------------------------------------
    # Charging control (called from ROS 2 service callbacks)
    # ------------------------------------------------------------------

    def start_draining(self):
        """Begin discharging."""
        self._is_depleted = False
        self._is_draining = True
        self._is_charging = False

    def stop_draining(self):
        """Pause discharging."""
        self._is_draining = False

    def start_charging(self):
        """Begin recharging."""
        self._is_charging = True
        self._is_draining = False

    def stop_charging(self):
        """Stop recharging."""
        self._is_charging = False

    # ------------------------------------------------------------------
    # Update (called from ROS 2 timer on main thread)
    # ------------------------------------------------------------------

    def update(self, speed: float, dt: float):
        """Advance the battery model.

        Args:
            speed: Horizontal vehicle speed in m/s.
            dt:    Elapsed wall-clock time since last call (seconds).

        Returns:
            Updated state dict, or ``None`` if rate limit not reached.
        """
        if self._is_depleted and not self._is_charging:
            return self._state

        # ── Charging path ─────────────────────────────────────────────
        if self._is_charging:
            charge_rate = self._capacity / max(self._charging_time, 0.001)  # Ah/h
            self._charge = min(
                self._capacity,
                self._charge + charge_rate * (dt / 3600.0),
            )
            if self._charge >= self._capacity:
                self._is_charging = False
                self._is_depleted = False
            self._state = self._make_state()
            return self._state

        # ── Discharging path ──────────────────────────────────────────
        if not self._is_draining:
            return self._state

        if self._consumption_mode == "velocity_based":
            total_power = self._base_power_load + self._power_per_speed * speed
        else:  # "constant"
            total_power = self._base_power_load

        v_oc = self._e0 + self._e1 * (1.0 - self._charge / self._capacity)
        v_oc = max(v_oc, 0.01)

        i_instant = total_power / v_oc

        if self._tau > 0.0:
            self._smoothed_current += (dt / self._tau) * (i_instant - self._smoothed_current)
        else:
            self._smoothed_current = i_instant

        self._charge -= self._smoothed_current * (dt / 3600.0)
        self._charge = max(self._charge, 0.0)

        if self._charge <= 0.0:
            self._is_depleted = True
            self._is_draining = False

        self._state = self._make_state(total_power=total_power, v_oc=v_oc)
        return self._state

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _make_state(self, total_power: float = 0.0, v_oc: float = None) -> dict:
        if v_oc is None:
            v_oc = self._e0 + self._e1 * (1.0 - self._charge / self._capacity)

        v_terminal = v_oc - self._smoothed_current * self._resistance
        charge_fraction = max(0.0, min(1.0, self._charge / self._capacity))

        return {
            "voltage":         float(v_terminal),
            "open_circuit_v":  float(v_oc),
            "current":         float(self._smoothed_current),
            "charge":          float(self._charge),
            "charge_fraction": float(charge_fraction),
            "percentage":      float(charge_fraction * 100.0),
            "power_load":      float(total_power),
            "is_depleted":     self._is_depleted,
        }
